src/ws/cockpitbranding.c

```diff
@@ -0,0 +1,278 @@
+  if (env && env[0])
+    return g_get_system_data_dirs ();
+
+  return NULL;
+}
+
+static void
+add_system_dirs (GPtrArray *dirs)
+{
+  const gchar * const* system;
+  system = get_system_data_dirs ();
+  while (system && system[0])
+    {
+      g_ptr_array_add (dirs, g_build_filename (system[0], "cockpit", "static", NULL));
+      system++;
+    }
+}
+
+gchar **
+cockpit_branding_calculate_static_roots (const gchar *os_id,
+                                         const gchar *os_variant_id,
+                                         gboolean is_local)
+{
+  GPtrArray *dirs;
+  gchar **roots;
+
+  dirs = g_ptr_array_new_with_free_func (g_free);
+
+  if (is_local)
+    add_system_dirs (dirs);
+
+  if (os_id)
+    {
+      if (os_variant_id)
+          g_ptr_array_add (dirs, g_strdup_printf (DATADIR "/cockpit/branding/%s-%s", os_id, os_variant_id));
+      g_ptr_array_add (dirs, g_strdup_printf (DATADIR "/cockpit/branding/%s", os_id));
+    }
+
+  if (!is_local)
+    add_system_dirs (dirs);
+
+  g_ptr_array_add (dirs, g_strdup (DATADIR "/cockpit/branding/default"));
+  g_ptr_array_add (dirs, g_strdup (DATADIR "/cockpit/static"));
+  g_ptr_array_add (dirs, NULL);
+
+  roots = cockpit_web_server_resolve_roots ((const gchar **)dirs->pdata);
+
+  g_ptr_array_free (dirs, TRUE);
+  return roots;
+}
+
+static void
+serve_branding_css_file (CockpitWebResponse *response,
+                         const gchar *path,
+                         const gchar **roots,
+                         GHashTable *os_release)
+{
+  if (os_release)
+    cockpit_web_response_css_template (response, path, roots, os_release);
+  else
+    cockpit_web_response_file (response, path, (const gchar **)roots);
+}
+
+typedef struct {
+  const gchar *path;
+  gchar *channel;
+
+  CockpitWebResponse *response;
+  CockpitWebService *service;
+
+  CockpitTransport *transport;
+  gulong transport_control;
+  gulong transport_closed;
+
+  GHashTable *local_os_release;
+
+} CockpitBrandingData;
+
+static void
+cockpit_branding_data_free (CockpitBrandingData *data)
+{
+  if (data->transport_control)
+    g_signal_handler_disconnect (data->transport, data->transport_control);
+  data->transport_control = 0;
+
+  if (data->transport_closed)
+    g_signal_handler_disconnect (data->transport, data->transport_closed);
+  data->transport_closed = 0;
+
+  g_object_unref (data->response);
+  g_object_unref (data->service);
+
+  if (data->local_os_release)
+    g_hash_table_unref (data->local_os_release);
+
+  g_free (data->channel);
+  g_free (data);
+}
+
+static void
+on_init_ready (CockpitBrandingData *data)
+{
+  gchar **roots = NULL;
+  /* owned by cockpit webservice */
+  GHashTable *os_release = cockpit_web_service_get_transport_os_release (data->service,
+                                                                         data->transport);
+
+  if (os_release)
+    {
+      roots = cockpit_branding_calculate_static_roots (g_hash_table_lookup (os_release, "ID"),
```

Comment:
It seems wasteful to recalculate this each time? Is there any way we can associate this with the transport? Perhaps using ````g_object_set_data_full()```